pilot/pkg/serviceregistry/kube/controller.go

```diff
@@ -435,6 +438,13 @@ func (c *Controller) InstancesByPort(hostname model.Hostname, reqSvcPort int,
 		return nil, nil
 	}
 
+	c.RLock()
```

Comment:
kill all the logic in the AppendInstanceHandler below.. It will cause unnecessary config reloads and mess with incremental EDs we have.

Instead do the following here:
``` if service.type == meshExternal {
   getKubeService(service.Hostname) // orwhatever way you get the kube service
  return []*ServiceInstance{
    {
        Endpoint: model.NetworkEndpoint{
				Address:     k8sSvc.Spec.ExternalName,
				Port:        portEntry.Port,
				ServicePort: portEntry,
			},
			Service: svc,
			Labels:  k8sSvc.Labels,
    }
  }
}
```

The reason to do so is one of performance. CDS code will call out to this function for all external services. And you could directly return service instances for such services. the Service itself is already monitored byother handlers. So anytime the service changes, we will be able to get the updated set of instances.

Reply:
That was my original approach. Didn't work.

Without adding a handler in `AppendInstanceHandler`, requests to ExternalName services don't get through (if made via the mesh). 

Looks like `AppendInstanceHandler`'s role is to actively propagate changes to the routing mesh whenever such changes occur in the backend model (k8s).